source/secure_tunneling_operations.c

```diff
@@ -0,0 +1,670 @@
+            level,
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: aws_secure_tunnel_options_storage http proxy host name set to " PRInSTR,
+            (void *)options_storage,
+            AWS_BYTE_CURSOR_PRI(options_storage->http_proxy_options.host));
+
+        AWS_LOGUF(
+            log_handle,
+            level,
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: aws_secure_tunnel_options_storage http proxy port set to %" PRIu16,
+            (void *)options_storage,
+            options_storage->http_proxy_options.port);
+
+        if (options_storage->http_proxy_options.proxy_strategy != NULL) {
+            AWS_LOGUF(
+                log_handle,
+                level,
+                AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+                "id=%p: aws_secure_tunnel_options_storage http proxy strategy set to (%p)",
+                (void *)options_storage,
+                (void *)options_storage->http_proxy_options.proxy_strategy);
+        }
+    }
+
+    bool is_service_id_used = false;
+
+    if (options_storage->service_id_1 != NULL) {
+        is_service_id_used = true;
+        AWS_LOGUF(
+            log_handle,
+            level,
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: aws_secure_tunnel_options_storage service id 1:%s",
+            (void *)options_storage,
+            aws_string_c_str(options_storage->service_id_1));
+    }
+
+    if (options_storage->service_id_2 != NULL) {
+        is_service_id_used = true;
+        AWS_LOGUF(
+            log_handle,
+            level,
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: aws_secure_tunnel_options_storage service id 2:%s",
+            (void *)options_storage,
+            aws_string_c_str(options_storage->service_id_2));
+    }
+
+    if (options_storage->service_id_3 != NULL) {
+        is_service_id_used = true;
+        AWS_LOGUF(
+            log_handle,
+            level,
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: aws_secure_tunnel_options_storage service id 3:%s",
+            (void *)options_storage,
+            aws_string_c_str(options_storage->service_id_3));
+    }
+
+    if (!is_service_id_used) {
+        AWS_LOGUF(
+            log_handle,
+            level,
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: aws_secure_tunnel_options_storage no service id set",
+            (void *)options_storage);
+    }
+}
+
+/*
+ * Clean up stored secure tunnel config
+ */
+void aws_secure_tunnel_options_storage_destroy(struct aws_secure_tunnel_options_storage *storage) {
+    if (storage == NULL) {
+        return;
+    }
+
+    aws_client_bootstrap_release(storage->bootstrap);
+    aws_http_proxy_config_destroy(storage->http_proxy_config);
+    aws_string_destroy(storage->endpoint_host);
+    aws_string_destroy(storage->access_token);
+    aws_string_destroy(storage->client_token);
+    aws_string_destroy(storage->service_id_1);
```

Comment:
Not comfortable with the fact that service ids are both configuration and dynamically modified.

Reply:
Service ids will only be configured/set on the receipt of a message from the secure tunnel service that gets send from them as soon as a websocket upgrade is accepted. This is on initial connect as well as on discon/recon. Not 100% what you mean by configuration but nobody sets these values outside of that message from the service.